### DISCUSSION TITLE
Polyphony troubles

### DISCUSSION THREAD

**Discussion by Slider2732** [marcel-licence/esp32_midi_sampler#13] [Q&A]
After changing various bits of code, I decided to rewind back to the original code as Marcel has presented and sort out polyphony. Also, to use something other than an Arduino MIDI interface, a real MIDI enabled keyboard instead. 
Something was messed up. Single notes were fine, chords dropped notes.
Setup now, is the original code running on an Audio Kit V2.2, triggered by my Yamaha YS200. Various MIDI Out channels have been tried, to test that functionality too.
Samples are 44.1K, 16 bit, downloaded from free sites such as:
https://freewavesamples.com/sample-type/bass/synth   <-- superb resource.

The same problem persists - single notes play perfectly and at whatever speed I press the keys. But chords often miss notes. The first chord will work fine, subsequent chords may or may not play all of their notes, it seems somewhat random. Usually only 2 notes will play of 4, for example.  
If time is given, such as 3 or 4 seconds, a chord will play fine, but not if pressed soon after the first chord.
Any sample does the same thing, from very short basses, to ultra long pads (all wavs are between 100K and ~1MB on the SD card).
Tried lowering the sample rate in the setup area of the code to 22050, no change. 
Tried uploading as a LOLIN D32 PRO because they have 8M PSRAM like the Audio Kit V2.2, no change.
Tried sternly warning it to behave after drinking several beers, no change  :)

So posted here for ideas on what's happening.

Is everyone elses polyphony fine?
If there are no such troubles, what may be the issue?

**Comment by marcel-licence** [comment]
Hi, sounds strange. Would you like to try to add some debug output into the NoteOn/NoteOff function. It may be good to know if the MIDI messages are not triggering the functions or otherwise the sampler itself exceeds its max polyphony.
You can also add in "Sampler_NoteoOn" in "       if (freePlayer != NULL)" an else path to show if no free instance was found. I am also working currently on this project. Thanks for sharing your observations

**Comment by Slider2732** [comment]
It didn't seem to change performance when the line above was added to the sampler_module.
Will look at the NoteOn, NoteOff with debug info  :)

Here's a quick example of what happens... must be easier to diagnose when heard.
I press 3 keys together (D,F,A), then repeat the chord over and over at 2 different speeds. Finally, a gap of about 3 seconds and the chord plays fine again. 
https://gofile.io/d/WNVclW

**Reply by marcel-licence** [reply]
Hm sounds like it should to me (as its implemented currently). The problem may be caused by the long release time (long sample) and that samples wont stopped to play new ones if polyphony has been exceeded. You could try to change the release time using Sampler_SetADSR_Release with another value (0: -> no release, 1: max release).

There is a mechanism implemented to find the quietest note which will be calculated using the adsr. So I think using release with a value < 1 (e.g. 0.9) will change the behavior. I think in your case triggering a new note when polyphony has been exceeded did grab always the same note or more randomly some other note.

**Comment by Slider2732** [comment]
That makes a lot of sense!
Thanks.
Yes, the effect does sound like a clipping of notes because of a long ADSR R section. Anyone with a controller would only need to move the corresponding slider. 
Also, the usual keyboard I'm using it with isn't velocity sensitive, so every note is at 127. 

...
Just tried your suggestion and there is some improvement.
Down to about 1.5 seconds between chords to play all notes. 90's Italian house piano lines will have to be a thing for the future lol 
Great point about the actual sample release time, especially for a piano.

**Reply by marcel-licence** [reply]
You could try to add setting the release when loading a sample. Velocity can be still remain 127. That shouldn't make any difference.
House sounds great =)

**Comment by Slider2732** [comment]
Update - by reducing the following line enough there's full polyphony (up to 8)
newPatch->release = 0.4f;  
Samples play in full when a key is pressed, but there's zero release time. It seems to work like an on/off.
Am going to use my Zoom FX stomp box. Did try to add reverb from within your code with a hard coded CC change, but it apparently locked out playback. Was likely retriggering it every millisecond.  Am more than happy to have to use the Zoom :) 

This is all stripped out and takes away a lot of features of your project, I understand that. Thanks for all the help.
Also, whatever the cycle count is for uploading sketches to these boards is impressing me. Nano's blow up around 200.

And, well done indeed on the sampler upgrades in your latest video!
The buttons are fixed and functional, a screen is shown running... awesome   :D
For screens, the SSD1306 for I2C and ST7735(s) for SPI are popular and cheap.

**Reply by marcel-licence** [reply]
Sounds good.
I can imagine that there is also a way to work with release = 1.0f with a small modification to allow getting the priority right.

I have some ST7725(s) here =) Only the SSD1306 is missing. I have a TTGO ESP32 LoRa Kit. There is a SSD1306 display installed.

200 uploads is the limit? That is really bad. I know that STM32 seems to be very durable. Regarding the ESP32 it's using external flash. It can be also used for storing data. Hopefully the flash will not die

**Comment by Slider2732** [comment]
Am presuming that nearly every MIDI controller with a keyboard has velocity sensitivity. 
The newPatch->release statement is where things are interesting without velocity, where it doesn't produce cut downs of the release time but instead enables or disables the full length of a sample.  
All I know is that I pressed 7 keys rapidly and every one of them played consistently :) 
In practice, with the Zoom MS-70CDR, the abrupt stops produced very audible glitches with reverb/echo. A route would seem to be to create a short tail using the sample value at cutoff, which decays using your inbuilt reverb. Trying that with fake CC message changes resulted in no output from the DAC. Am hoping to do such a thing though to give sounds a more natural cut off.

For priority, it may be easiest to put the matter to the ears of the musician and remove the problem. If cutting and glitches happen, people would know their ultra long decays are causing it and know it as a normal limitation. Personally, I would expect 5x 1.5 second piano samples to play, but only 3x 6 seconds long vocal Aah's. I mean, there's an easy logic for a sampler's abilities.
   
The screen and board mod are definitely going to happen here!
Will be using 1/4W 1.8K resistors and will put a swearing jar at my side haha. 
I bought 5x I2C SSD1306's for $17 (USA) on seeing Erich's comment about rejigging the I2C bus. RXD2 was put on IO3 as Serial1 and it partially worked with the Adafruit library. It cut out displaying messages when the AC101 setup ran, due to IO21, I think. The next day, your vid showed a relatively pin hungry SPI display running!
Am just thinking that such displays as the ST7735 and SSD1306 are what most people would have, rather than the capable but expensive Nextion's, as you pointed out. 
An analog input was also in the description of the vid - yep, ranges determining functions is a useful feature. Same sort of thing as the wave type selection in the synth. Very useful if the screen showed the info. 

Oh, the Nano's have been a mess over the years. It's normally not the flash that goes, but the diode underneath the board. Various manufacturers, same problem. In my experience, they blow if uploads are made rapidly and the cycle count is around 200 - I have completely no idea why. There is a fix, to run a wire from pin 1 of the USB input to where the output of the diode should go. The pads tend to fall off if trying to replace the diode. 
Otherwise, larger cycle counts can start to introduce upload errors. The sort of thing most don't see, because they buy the parts for an Instructables project lol and they're only a couple of $. People say the count is in the thousands, but probably wisely buy genuine Arduino's. 
STM32F103's have been good yep, for price and size. A bit odd to upload to, but capable.

^ sorry for the Sunday ramble

**Reply by marcel-licence** [reply]
The Arduino Nano was the reason I've never used arduino before. Because I used AVR's with other tools on breadboard and they were cheap. The arduino was too expensive and just the same microcontroller.

From past I've also some STM32 discovery boards. Very powerful and about 10 ~%^ each. 

Yes the ST7735 is pin hungry. I try to reduce the required pin count to have SPI and I2C available the same time. 

I also observed a click at the end of the playback of samples. I also want to get it fixed. That sould not happen. Tweaking the adsr is actually a possible workaround
